### PR TITLE
[TEST] Clean surface fixtures

### LINKED ISSUE #4638: Clean-up pytest fixtures for creating sample surface images/masks
Yes it does look like we need better fixtures for creating sample surface images.

I was super confused as well because I could not import any of the existing fixtures for making sample surface masks. 

Then I noticed the same thing with sample surface images, and there was an existing `make_mini_img` being used repeatedly. So I created a similar one for masks.

_Originally posted by @man-shu in https://github.com/nilearn/nilearn/pull/4630#discussion_r1799772743_

---

## TODOs

- [x] make sure that surface related contain `surf` to easily distinguish them from the volume one:
- [ ] when cleaning the test suite we should have a fixture that creates very small gifti files to disk to test this with rather than relying on datasets.
  _Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/pull/4589#discussion_r1799843886_
            

--- 

- closes none but relates to #4568

### PR DESCRIPTION
**Pull request by man-shu**
- Closes #4638

Changes proposed in this pull request:

- remove unnecessary fixtures
- update tests to use new fixtures

### CODE REVIEW

File: nilearn/experimental/surface/tests/test_surface_image.py [hunk 1]
```diff
@@ -25,22 +26,23 @@ def test_compare_file_and_inmemory_mesh(mini_mesh, tmp_path):
...
-    img = make_mini_img()
-    assert img.shape == (9,)
-    img = make_mini_img((3,))
-    assert img.shape == (3, 9)
-    img = make_mini_img((7, 3))
-    assert img.shape == (7, 3, 9)
+@pytest.mark.xfail(reason="Parameterizing the new fixture is not working.")
+@pytest.mark.parametrize(
+    "make_surface_img", [(1,), (3,), (7, 3)], indirect=True
+)
```
Comment (man-shu) [inline nilearn/experimental/surface/tests/test_surface_image.py:32]: Parameterizing the new fixtures is not super elegant and probably confusing

File: nilearn/experimental/surface/tests/test_surface_image.py [hunk 2]
```diff
@@ -25,22 +26,23 @@ def test_compare_file_and_inmemory_mesh(mini_mesh, tmp_path):
     assert np.array_equal(left.faces, left_loaded.faces)
 
 
-def test_surface_image_shape(make_mini_img):
-    img = make_mini_img()
-    assert img.shape == (9,)
-    img = make_mini_img((3,))
-    assert img.shape == (3, 9)
-    img = make_mini_img((7, 3))
-    assert img.shape == (7, 3, 9)
```
Comment (man-shu) [inline nilearn/experimental/surface/tests/test_surface_image.py:34]: What I don't understand is why the data in SurfaceImage would ever be 3D.

AFAIU, the shape is supposed to be
 - `(1, n_vertices)` when there is only one image 
 - `(n_timepoints, n_vertices)` when we have timeseries surface image

File: nilearn/experimental/surface/tests/test_surface_image.py (hunk 1 above)
Comment (man-shu) [reply nilearn/experimental/surface/tests/test_surface_image.py]: Ok resolved this one with @Remi-Gau 's help

File: nilearn/experimental/surface/tests/test_surface_image.py (hunk 2 above)
Comment (man-shu) [reply nilearn/experimental/surface/tests/test_surface_image.py:34]: So @Remi-Gau mentioned 3D arrays might be possible with layer fMRI, where you can sample a given vertex at different cortical depths at a given time point.

However, the fact if we wanna support this still open for discussion

File: nilearn/experimental/surface/tests/test_maskers.py [hunk 3]
```diff
@@ -77,15 +89,24 @@ def test_transform_inverse_transform(shape, make_mini_img, assert_img_equal):
...
-    img = make_mini_img(shape)
-    masker = SurfaceMasker(mini_mask).fit(img)
+    img = make_surface_img(shape)
+    mask = make_surface_mask(n_zeros_in_mask)
+    masker = SurfaceMasker(mask).fit(img)
     masked_img = masker.transform(img)
-    assert masked_img.shape == (*shape, img.shape[-1] - 2)
-    assert np.array_equal(masked_img.ravel()[:7], [2, 3, 4, 20, 30, 40, 50.0])
+    assert masked_img.shape == (*shape, img.shape[-1] - n_zeros_in_mask)
+    assert np.array_equal(masked_img.ravel()[:7], [2, 3, 4, 20, 30, 40, 50])
```
Comment (man-shu) [inline nilearn/experimental/surface/tests/test_maskers.py:109]: Just a note for myself:

We are comparing hard-coded array values here. 
So, as discussed with @Remi-Gau, it might be best to define new SurfaceImage with specific data instead of using the fixture.

File: nilearn/experimental/surface/tests/test_maskers.py [hunk 4]
```diff
@@ -11,62 +11,74 @@
...
 @pytest.mark.parametrize("shape", [(), (1,), (3,), (3, 2)])
-def test_transform_inverse_transform(shape, make_mini_img, assert_img_equal):
-    img = make_mini_img(shape)
+def test_transform_inverse_transform(
+    make_surface_img, shape, assert_img_equal
+):
+    img = make_surface_img(shape)
     masker = SurfaceMasker().fit(img)
     masked_img = masker.transform(img)
     assert np.array_equal(
```
Comment (man-shu) [inline nilearn/experimental/surface/tests/test_maskers.py:84]: Again comparing hard-coded values. So same here

File: nilearn/experimental/surface/tests/test_maskers.py [hunk 5]
```diff
@@ -79,14 +98,35 @@ def test_transform_inverse_transform(shape, make_mini_img, assert_img_equal):
...
-    masker = SurfaceMasker(mini_mask).fit(img)
+    # make a sample image with data 1-4 on left part
+    # and 10-50 on right part
+    mesh = surf_mesh()
+    img_data = {}
+    for i, (key, val) in enumerate(mesh.parts.items()):
+        data_shape = (*shape, val.n_vertices)
+        data_part = (
+            np.arange(np.prod(data_shape)).reshape(data_shape) + 1.0
+        ) * 10**i
```
Comment (bthirion) [inline nilearn/experimental/surface/tests/test_maskers.py:111]: I don't see why the values would be 1-4 and 10^-50 then ?

File: nilearn/experimental/surface/tests/test_surface_image.py (hunk 2 above)
Comment (bthirion) [reply nilearn/experimental/surface/tests/test_surface_image.py:34]: ATM we're not supporting it. This should be discussed in a coredev meeting, I have no experience with layered fMRI on my side.

**Review by bthirion — COMMENTED** [review]
We probably have to agree on whether surface data can be more than 2D.

File: nilearn/experimental/surface/tests/test_surface_image.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/experimental/surface/tests/test_surface_image.py:34]: My position on this would be: we were not supporting it, so let's not add it just yet.

I would revisit the issues when layer MRI user ask for it.

File: nilearn/experimental/surface/tests/test_maskers.py (hunk 5 above)
Comment (man-shu) [reply nilearn/experimental/surface/tests/test_maskers.py:111]: It's not 10^-50, but between 10 to 50. But it would be easier to read if I just hard-code the values explicitly.

Comment (man-shu) [reply nilearn/experimental/surface/tests/test_maskers.py:111]: Ok yes just the first sample would be between 1-4 (on left) and 10-50 (on right).

Comment (man-shu) [reply nilearn/experimental/surface/tests/test_maskers.py:111]: updated the comment https://github.com/nilearn/nilearn/pull/4657/commits/bb9e9b1d0836f6ba6441e7876e79e33de366806b

**Review by bthirion — APPROVED** [review]
LGTM.

**Comment by Remi-Gau** [comment]
@man-shu 
As discussed, I Will move the fixtures into the main conftest.py

**Comment by man-shu** [comment]
Thanks for taking care of this, @Remi-Gau !

File: nilearn/conftest.py [hunk 6]
```diff
@@ -377,3 +371,154 @@ def img_atlas(shape_3d_default, affine_mni):
...
+    """Flip hemispheres of a surface image."""
+
+    def f(img):
+        return SurfaceImage(flip(img.mesh), flip(img.data))
+
+    return f
+
+
+@pytest.fixture
+def assert_img_equal():
```
Comment (Remi-Gau) [inline nilearn/conftest.py:502]: rename to assert_surf_img_equal

File: nilearn/conftest.py [hunk 7]
```diff
@@ -377,3 +371,154 @@ def img_atlas(shape_3d_default, affine_mni):
...
+    def f(img_1, img_2):
+        assert set(img_1.data.parts.keys()) == set(img_2.data.parts.keys())
+        for key in img_1.data.parts:
+            assert np.array_equal(img_1.data.parts[key], img_2.data.parts[key])
+
+    return f
+
+
+@pytest.fixture
+def drop_img_part():
```
Comment (Remi-Gau) [inline nilearn/conftest.py:514]: rename to drop_surf_img_part

File: nilearn/conftest.py [hunk 8]
```diff
@@ -377,3 +371,154 @@ def img_atlas(shape_3d_default, affine_mni):
...
+    def f(poly_obj):
+        keys = list(poly_obj.parts.keys())
+        keys = [keys[-1]] + keys[:-1]
+        return dict(zip(keys, poly_obj.parts.values()))
+
+    return f
+
+
+@pytest.fixture
+def flip_img(flip):
```
Comment (Remi-Gau) [inline nilearn/conftest.py:492]: flip_surf_img

File: nilearn/conftest.py [hunk 9]
```diff
@@ -377,3 +371,154 @@ def img_atlas(shape_3d_default, affine_mni):
...
+            "right": np.asarray([1, 1, 0, 0, 0]),
+        }
+
+        return SurfaceImage(mesh, data)
+
+    return _surface_label_img
+
+
+@pytest.fixture
+def flip():
```
Comment (Remi-Gau) [inline nilearn/conftest.py:480]: flip_parts

File: nilearn/decoding/tests/test_decoder.py [hunk 10]
```diff
@@ -1088,65 +1088,53 @@ def test_decoder_decision_function_raises_value_error(
...
+    img = surf_img(shape)
+    y = rng.random(shape)
+    return img, y
 
 
+@pytest.mark.filterwarnings("ignore:Overriding provided")
 def test_decoder_apply_mask_surface(_make_surface_class_data):
     """Test _apply_mask on surface image."""
-    X, _ = _make_surface_class_data()
+    X, y = _make_surface_class_data
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1109]:
```suggestion
    X, _ = _make_surface_class_data
```

File: nilearn/decoding/tests/test_decoder.py [hunk 11]
```diff
@@ -1174,22 +1165,24 @@ def test_decoder_adjust_screening_lessthan_mask_surface(
...
+    surf_mask,
+    _make_surface_class_data,
+    screening_percentile=80,
 ):
     """When mask size is greater than screening percentile wrt to the mesh
     size, it is changed to 100% of mask.
     """
-    mask = _make_surface_mask()
-    img, y = _make_surface_class_data()
+    mask = surf_mask()
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1184]: no need to reassign the fixture

File: nilearn/decoding/tests/test_decoder.py [hunk 12]
```diff
@@ -1211,33 +1203,37 @@ def test_decoder_adjust_screening_greaterthan_mask_surface(
...
 def test_decoder_fit_surface_with_mask_image(
-    _make_surface_class_data, decoder, mini_mask
+    _make_surface_class_data, decoder, surf_mask
 ):
     """Test fit for surface image."""
     warnings.simplefilter("ignore", ConvergenceWarning)
-    X, y = _make_surface_class_data()
-    model = decoder(mask=mini_mask)
+    X, y = _make_surface_class_data
+    mask = surf_mask()
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1221]:
```suggestion
    mask = surf_mask()
```

no need to reassign the fixture

File: nilearn/decoding/tests/test_decoder.py [hunk 13]
```diff
@@ -1211,33 +1203,37 @@ def test_decoder_adjust_screening_greaterthan_mask_surface(
...
+@pytest.mark.filterwarnings("ignore:Overriding provided")
 @pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
 def test_decoder_error_incompatible_surface_mask_and_volume_data(
-    decoder, mini_mask, tiny_binary_classification_data
+    decoder, surf_mask, tiny_binary_classification_data
 ):
     """Test error when fitting volume data with a surface mask."""
     data_volume, y, _ = tiny_binary_classification_data
-    model = decoder(mask=mini_mask)
+    mask = surf_mask()
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1235]: no need to reassign the fixture

File: nilearn/experimental/reporting/tests/test_html_report.py [hunk 14]
```diff
@@ -1,38 +1,43 @@
...
-@pytest.mark.parametrize("mask_img", [None, return_mini_binary_mask()])
-def test_surface_masker_minimal_report_fit(mask_img, mini_img, reports):
+@pytest.mark.parametrize("empty_mask", [True, False])
+def test_surface_masker_minimal_report_fit(
+    surf_mask, empty_mask, surf_img, reports
+):
     """Test minimal report generation with fit."""
-    masker = SurfaceMasker(mask_img, reports=reports)
-    masker.fit_transform(mini_img)
+    mask = surf_mask(empty=empty_mask)
```
Comment (Remi-Gau) [inline nilearn/experimental/reporting/tests/test_html_report.py:25]: pass it directly

File: nilearn/experimental/reporting/tests/test_html_report.py [hunk 15]
```diff
@@ -1,38 +1,43 @@
...
+    masker = SurfaceMasker(mask, reports=reports)
+    img = surf_img()
+    masker.fit_transform(img)
```
Comment (Remi-Gau) [inline nilearn/experimental/reporting/tests/test_html_report.py:28]: pass it directly

File: nilearn/experimental/reporting/tests/test_html_report.py [hunk 16]
```diff
@@ -1,38 +1,43 @@
...
     assert '<div class="image">' in str(report)
 
 
-def test_surface_masker_report_no_report(mini_img):
+def test_surface_masker_report_no_report(surf_img):
     """Check content of no report."""
     masker = SurfaceMasker(reports=False)
-    masker.fit_transform(mini_img)
+    img = surf_img()
+    masker.fit_transform(img)
```
Comment (Remi-Gau) [inline nilearn/experimental/reporting/tests/test_html_report.py:40]: pass it directly

File: nilearn/experimental/reporting/tests/test_html_report.py [hunk 17]
```diff
@@ -54,9 +60,10 @@ def test_surface_label_masker_report_unfitted(
     assert "Make sure to run `fit`" not in str(report)
 
 
-def test_surface_label_masker_report_no_report(mini_label_img):
+def test_surface_label_masker_report_no_report(surf_label_img):
     """Check content of no report."""
-    masker = SurfaceLabelsMasker(mini_label_img, reports=False)
+    label_img = surf_label_img()
+    masker = SurfaceLabelsMasker(label_img, reports=False)
```
Comment (Remi-Gau) [inline nilearn/experimental/reporting/tests/test_html_report.py:66]: pass it directly

File: nilearn/experimental/reporting/tests/test_html_report.py [hunk 18]
```diff
@@ -43,9 +48,10 @@ def test_surface_masker_report_no_report(mini_img):
 @pytest.mark.parametrize("reports", [True, False])
 @pytest.mark.parametrize("label_names", [None, ["region 1", "region 2"]])
 def test_surface_label_masker_report_unfitted(
-    mini_label_img, label_names, reports
+    surf_label_img, label_names, reports
 ):
-    masker = SurfaceLabelsMasker(mini_label_img, label_names, reports=reports)
+    label_img = surf_label_img()
```
Comment (Remi-Gau) [inline nilearn/experimental/reporting/tests/test_html_report.py:53]: pass it directly

File: nilearn/conftest.py [hunk 19]
```diff
@@ -377,3 +371,167 @@ def img_atlas(shape_3d_default, affine_mni):
...
+            "right": np.asarray([1, 1, 0, 0, 0]),
+        }
+
+        return SurfaceImage(mesh, data)
+
+    return _surface_label_img
+
+
+@pytest.fixture
+def flip_surf_img_parts():
```
Comment (Remi-Gau) [inline nilearn/conftest.py:480]: It still feels weird to me to use the last four as fixture and not as functions that we store somewhere in nilearn._utils ? Maybe nilearn._utils.testing?

But I don't think this is urgent. Let's open a side issue for these and move on to merge.

**Review by Remi-Gau — APPROVED** [review]

File: nilearn/conftest.py (hunk 19 above)
Comment (Remi-Gau) [reply nilearn/conftest.py:480]: probably something to do as we try to also better fixturize test for the old surface code
